Add HBV model

buzacott opened the pull request:
### All Submissions:

* [X] Have you followed the guidelines in our **CONTRIBUTING** document?
* [X] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

<!-- Does this resolve or supercede an open Issue or Pull Request?
If so, write a line like Closes #12345 for each Issue or PR number that should be closed
if this Pull Request is merged. -->

Closes #46 

<!-- You can erase any parts of this template not applicable to your Pull Request. -->


<!-- These have been added according to the rOpenSci guidelines. -->
<!-- From: https://www.talater.com/open-source-templates/#/page/99 -->

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue. List the bug ID if applicable)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing core functionality to change) 


### New Feature Submissions checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
1. [X] Have you ensured there are no conflicts with major packages (e.g. tidyverse, MASS) or base R packages?
2. [X] Does your new feature require documentation, and if so, have you updated the documentation?
3. [ ] Have you added tests for your features, and do your features pass all tests?
4. [X] Have you linted your code with lintr locally prior to submission?
5. [X] Have you styled your code with styler locally prior to submission?
6. [ ] Have you [benchmarked](https://www.alexejgossmann.com/benchmarking_r/) the performance of your code if applicable (e.g. with microbenchmark)?
7. [X] Does your new feature comply with existing model APIs (e.g. for models, optimisation algorithms, objective functions)?


### Changes to Core Features:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
* [X] Have you added an explanation of what your changes do and why you'd like us to include them?
* [X] Have you added tests for your core features and do your core features pass all tests?
* [X] Have you successfully run tests with your changes locally?
* [X] Does your code run on all major platforms (Windows, macOS, Linux)?
* [ ] Have you demonstrated backwards compatibility and compatibility with the current development version, or discussed a potential break in backwards compatibility?
* [ ] Have you updated the package vignettes? 

This PR adds the HBV model to `hydromad`. The HBV model implemented is based off HBV light as described in https://doi.org/10.5194/hess-16-3315-2012. This branch is built off the KGE branch so that PR should be merged first.

Summary of changes
* Adds HBV model in R and C++
* Adds Corin test dataset as daily mean air temperature is needed
* Added HBV test to the testthat suite

buzacott commented:
Tests finally seem to be passing.

One area I could take your input @josephguillaume is whether the groundwater routine should be moved to `hbvrouting`.  At the moment in `hbvrouting` there is only the triangular routing function and strictly `hbv` isn't just doing SMA. It wouldn't be hard to move if you think the change makes it more consistent with the rest of the package

josephguillaume commented on R/Corin.R:
```diff
@@ -0,0 +1,32 @@
+#' * `T` (daily average temperature, ºC)
+#' @details
+#' * Daily rainfall at Cotter Hut (ID: 570946).
+#' * Daily mean streamflow at Cotter River at Gingera (ID: 410730).
```
This this seems to be the same gauge as ?Cotter
I'm happy merging this for the time being, but before the 1.0 release we need to rationalise the datasets included #72

buzacott replied:
Indeed it is the same gauge. I didn't want to modify the existing dataset in case it broke something for the time being, but I can revisit this. 

Maybe adding temperature data to the existing Cotter dataset would be easier.

josephguillaume replied:
Yep, fine like this.
I've added a comment to #72 . It'll be good to get your input into that when the time comes

josephguillaume commented on man/hbv.Rd:
```diff
@@ -0,0 +1,135 @@
+snowpack (sp), water content of snowpack (wc), soil moisture (sm), actual
+evapotranspiration (ETa), upper groundwater storage (uz), lower groundwater
+storage (lz).
+}
```
Could you please add a comment about the source/reasoning behind values for default parameters and parameter ranges?

josephguillaume commented on man/hbv.Rd:
```diff
@@ -0,0 +1,135 @@
+data is required for the snow routine. Daily potential evapotranspiration
+(PET) data is required as the routine to calculate daily PET using long term
+mean PET and daily temperature is not included.
+
```
At some point we'll need to check whether this snow routing is equivalent to the one in snow.sim, and work out how we can harmonise how snow is dealt with across all models #24

josephguillaume commented on tests/testthat/test-sma-sim.R:
```diff
@@ -133,6 +133,20 @@ test_that("sacramento simulation is the same in R and C", {
   }
 })
 
+test_that("hbv simulation is the same in R and C", {
```
Sorry! Could you please modify this to use the new patrick test structure?

josephguillaume commented on tests/testthat/test-sma-sim.R:
```diff
@@ -133,6 +133,20 @@ test_that("sacramento simulation is the same in R and C", {
 
+test_that("hbv simulation is the same in R and C", {
+  set.seed(0)
+  hbvDATA <- cbind(P = P, E = E, T = E - 15)
```
What's the motivation behind T = E -15?

buzacott replied:
E appeared to be temperature data, so I used E-15 so that temperatures went below 0 to invoke the snow routine. It appears I did not think of dealing with E being temperature and not PET for the model. Got a bit lazy.

I'll check out the patrick tests

josephguillaume replied:
Ah. 
Let's just use the existing data for now, but with DATATE having E in mm rather than temperature, even if this means test coverage won't be 100%.
Later:
Let's add a separate test dataset for snow #163 
Let's also separately decide how to handle T vs E #32

josephguillaume commented on man/hbv.Rd:
```diff
@@ -0,0 +1,135 @@
+}
+\arguments{
+\item{DATA}{A time-series like object with columns P (precipitation in mm),
+E (potential evapotranspiration in mm), T (average air temperature in ºC).}
```
The separation of E and T is new in hydromad. I think it makes sense but we'll need to think about how this affects, e.g. snow.sim before the 1.0 release

josephguillaume commented:
> One area I could take your input @josephguillaume is whether the groundwater routine should be moved to `hbvrouting`. At the moment in `hbvrouting` there is only the triangular routing function and strictly `hbv` isn't just doing SMA. It wouldn't be hard to move if you think the change makes it more consistent with the rest of the package

Both gr4j and gr4jrouting are a bit ambiguous too.
The split between sma and routing was driven by the use of the effective rainfall paradigm in IHACRES.
Is there a particular variable in HBV that has a similar interpretation?

josephguillaume commented:
It would be good if we could come up with a policy for verification of correctness of model implementation.
What did you do when writing this code? Is there anything we could include as an automated test, e.g. a sample input/output dataset to be included only for testing?

buzacott commented:
> It would be good if we could come up with a policy for verification of correctness of model implementation.
> What did you do when writing this code? Is there anything we could include as an automated test, e.g. a sample input/output dataset to be included only for testing?

I mainly used the paper by Seibert et al. to implement the model steps, however I did check out some other implementations. I haven't compared with other implementations as you suggest #46. There could be minor differences as there are a lot of implementations of HBV, but I agree a comparison would be good to see if the models are somewhat consistent

buzacott commented:
> > One area I could take your input @josephguillaume is whether the groundwater routine should be moved to `hbvrouting`. At the moment in `hbvrouting` there is only the triangular routing function and strictly `hbv` isn't just doing SMA. It wouldn't be hard to move if you think the change makes it more consistent with the rest of the package
> 
> Both gr4j and gr4jrouting are a bit ambiguous too.
> The split between sma and routing was driven by the use of the effective rainfall paradigm in IHACRES.
> Is there a particular variable in HBV that has a similar interpretation?

Effective rainfall is in the model was not explicitly defined, but `runoff + recharge` is the effective rainfall from each timestep and could be the net result of the SMA module. 

I've shifted it the GW routine to HBV routing, so the `hbv` is doing SMA and `hbvrouting` includes the GW routine and triangular weighting function.

josephguillaume commented on R/hbv.R:
```diff
@@ -0,0 +1,436 @@
+#' @param return_components Whether to return state variables of the routing
+#' routine.
+#' @details This implementation of this HBV model closely follows the
+#' description of HBV light by Seibert and Vis, 2009. Daily average temperature
```
Did you still want to do a check comparing with sample input/output from the HBVlight model?
We could add a disclaimer that direct compatibility is not guaranteed, but that might discourage people from using this implementation.

buzacott replied:
Yep this is something I still want to confirm. I suspect it isn't exact at the moment due to the description of the triangular routing routine and accepting non-integer values. I'll try and have a look this weekend.

josephguillaume replied:
Thanks! I am happy when you're happy - we just need to have a clear and convincing statement about what the status of this implementation is compared to other implementations of HBV.

buzacott commented:
@josephguillaume I'm happy with it now. I updated the code and did some checking against the output of the HBV-light application and the models are obtaining the same output now. I also added the HBV implementation to calculate PET using monthly/daily average PET and temperature values.

There is also now a vignette for the model with a basic overview of the module and there is a comparison of output using the Corin dataset and the HBV-land dataset from HBV-light.

Unrelated, but the test-coverage action is failing and it seems like there is some compiling error on one of the dependencies to setup the action. After a quick look it looks like it is a fortran/macOS issue, but perhaps the action should be run using linux as software compatibility will be more robust there.

josephguillaume commented:
Thanks! 
You note "the models are obtaining the same output now" and "there is a comparison of output using the Corin dataset and the HBV-land dataset from HBV-light".
It might be worth adding the model output comparison as a regression test, which then means we can request this also of future implementations that want to be included in hydromad. Let me know if you think that's not worthwhile.

josephguillaume commented:
A quick look of the vignette looks good to me. 
@WillemVervoort - could you please check you're happy with it too? (Just to have a second reviewer on vignettes)

WillemVervoort commented:
Small changes in the vignette to make it more compatible with the standard hydromad implementation

## Data requirements

Most of the required data for the model are similar to standard hydromad:

- P: timeseries of precipitation (mm)
- Q: timeseries of observed discharge (mm)
- E: timeseries of potential evapotranspiration (PET) 
Specifically for HBV, a further column needs to be added with:  
- T: timeseries of average temperature (ºC)


I am confused about this part about the options for PET

There are three options:
- A timeseries of the same length as the main data (I understand this, this is same as standard hydromad)
- A vector of length 365 with daily average PET for each day of the year: but this cannot be part of the PQE file, do you need to then have missing values in the PQE data
- A vector of length 12 with average daily PET for each month of the year
A vector of length 12 or 365 of mean temperature can also be provided for use in calculating PET. If it is not provided then mean monthly temperature is calculated from the timeseries of temperature.

Or does this relate to the point where PET is calculated using the mean temperature, i.e.:
If daily estimates of potential evapotranspiration are not available for your catchment, then they can be calculated using the HBV method (see equation 7 Seibert and Vis, 2012).

josephguillaume commented:
Thanks @buzacott !
As discussed, we'll add the regression tests in a separate PR.
